### PR TITLE
feat(context): add support for Pull Request in GitHub Context

### PR DESCRIPTION
**Pull request by zwpaper**

### CODE REVIEW

File: crates/tabby-common/src/api/structured_doc.rs [hunk 1]
```diff
@@ -65,6 +66,15 @@ pub struct DocSearchIssueDocument {
     pub closed: bool,
 }
 
+#[derive(Clone)]
+pub struct DocSearchPullRequest {
+    pub title: String,
+    pub link: String,
+    pub body: String,
+    pub patch: String,
```
Comment (wsxiaoys) [inline crates/tabby-common/src/api/structured_doc.rs:74]: Add document on this field is in patch / diff format (which standard, example, etc)

File: crates/tabby-common/src/api/structured_doc.rs [hunk 2]
```diff
@@ -65,6 +66,15 @@ pub struct DocSearchIssueDocument {
...
+    pub state: String,
```
Comment (wsxiaoys) [inline crates/tabby-common/src/api/structured_doc.rs:75]:
```suggestion
    pub closed: boolean,
```

Comment (zwpaper) [reply crates/tabby-common/src/api/structured_doc.rs:75]: but there are three state for PRs, Open, Closed, Merged

Comment (wsxiaoys) [reply crates/tabby-common/src/api/structured_doc.rs:75]: offline discussed

File: crates/tabby-common/src/api/structured_doc.rs [hunk 3]
```diff
@@ -20,6 +20,7 @@ pub struct DocSearchHit {
 pub enum DocSearchDocument {
     Web(DocSearchWebDocument),
     Issue(DocSearchIssueDocument),
+    Pull(DocSearchPullRequest),
```
Comment (wsxiaoys) [inline crates/tabby-common/src/api/structured_doc.rs:23]:
```suggestion
    PullRequest(DocSearchPullRequest),
```

Please also update other naming /variables to reflect the change.

File: crates/tabby-index/src/structured_doc/types/pull.rs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+            fields::pull::MERGED: self.merged,
+        })
+    }
+
+    async fn build_chunk_attributes(
+        &self,
+        embedding: Arc<dyn Embedding>,
+    ) -> BoxStream<JoinHandle<(Vec<String>, serde_json::Value)>> {
+        // currently not indexing the diff
+        let text = format!("{}\n\n{}", self.title, self.body);
```
Comment (wsxiaoys) [inline crates/tabby-index/src/structured_doc/types/pull.rs:47]: please refer how issues compute embeddings for both title and body, then merge them into one

File: crates/tabby-common/src/api/structured_doc.rs [hunk 5]
```diff
@@ -65,6 +66,15 @@ pub struct DocSearchIssueDocument {
     pub closed: bool,
 }
 
+#[derive(Clone)]
+pub struct DocSearchPullRequest {
```
Comment (wsxiaoys) [inline crates/tabby-common/src/api/structured_doc.rs:70]:
```suggestion
pub struct DocSearchPullDocument {
```

File: crates/tabby-index/src/structured_doc/types/pull.rs [hunk 6]
```diff
@@ -10,21 +10,21 @@ use tokio::task::JoinHandle;
...
-pub struct PullRequest {
+pub struct PullDocument {
     pub link: String,
     pub title: String,
     pub body: String,
+
+    /// Code changes can be represented in two formats: diff and patch.
+    /// Since commits are not relevant, the diff format is used.
+    /// For details on the diff format, refer to:
+    /// https://git-scm.com/docs/diff-format#_combined_diff_format
```
Comment (wsxiaoys) [inline crates/tabby-index/src/structured_doc/types/pull.rs:21]: You don't have to explain the reason to select "diff format". Directly explain what's stored and what spec is used would be sufficient.

File: ee/tabby-webserver/src/service/thread.rs [hunk 7]
```diff
@@ -108,6 +108,7 @@ impl ThreadService for ThreadServiceImpl {
             )
             .await?;
 
+        println!("assistant_message_id: {:?}", assistant_message_id);
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/service/thread.rs:111]:
```suggestion
```

**Review by wsxiaoys — COMMENTED** [review]
Otherwise LGTM

File: crates/tabby-index/src/structured_doc/types/pull.rs (hunk 6 above)
Comment (zwpaper) [reply crates/tabby-index/src/structured_doc/types/pull.rs]: comments updated

File: ee/tabby-webserver/src/service/background_job/third_party_integration/pulls.rs [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+                    Ok(x) if x.len() < 1024*1024 => x,
+                    Ok(_) => {
+                        logkit::warn!("Pull request {} diff is larger than 1MB, skipping", url);
+                        continue
+                    }
+                    Err(e) => {
+                        logkit::error!("Failed to fetch pull request patch for {}: {}", url, e);
+                        continue
+                    }
+                };
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/service/background_job/third_party_integration/pulls.rs:64]:
```suggestion
                    Err(e) => {
                        logkit::error!("Failed to fetch pull request diff for {}: {}", url, e);
                        continue
                    }
                };
```

File: ee/tabby-webserver/src/service/background_job/third_party_integration/pulls.rs [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+                    }
+                }
+
+                let url = pull.html_url.map(|url| url.to_string()).unwrap_or_else(|| pull.url);
+                let diff = match octocrab.pulls(&owner, &repo).get_diff(pull.number).await {
+                    Ok(x) if x.len() < 1024*1024 => x,
+                    Ok(_) => {
+                        logkit::warn!("Pull request {} diff is larger than 1MB, skipping", url);
+                        continue
+                    }
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/service/background_job/third_party_integration/pulls.rs:59]:
```suggestion
                    Ok(x) if x.len() < 1024*1024 * 10 => x,
                    Ok(_) => {
                        logkit::warn!("Pull request {} diff is larger than 10MB, skipping", url);
                        continue
                    }
```

**Pull request merged**